### PR TITLE
Feat - Programatically toggle live chat widget

### LINKED ISSUE #1721: Dynamic control of bubble visibility at runtime
Sometimes it comes handy to be able to hide/show the bubble depending on runtime conditions e.g. some modal dialog is shown thus hide the bubble. Another example would be having a dedicated button for toggling the widget at certain states of the app.

An idea would be adding to the sdk  something like window.$chatwoot.showBubble() , window.$chatwoot.hideBubble() as well as
a window.$chatwoot.bubbleIsShown flag to reflect on the current state of the bubble.

### PR DESCRIPTION
**Pull request by fayazara**
# Pull Request Template

## Description
Adds new sdk method `toggleBubbleVisibility ('hide'/'show')` which toggles the widget, this toggle witll also set the `hideMessageBubble` state value.

https://user-images.githubusercontent.com/15716057/159111463-f18ce475-bdea-49b7-b3f4-edafdcfdca87.mp4

Fixes #1721 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Tested locally on the `/widget_tests` page.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by fayazara** [comment]
@pranavrajs, does this approach work? Toggling the widget parent element via display property.

File: app/javascript/packs/sdk.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ const runSDK = ({ baseUrl, websiteToken }) => {
     hasLoaded: false,
     hideMessageBubble: chatwootSettings.hideMessageBubble || false,
     isOpen: false,
+    isWidgetVisible: true,
```
Comment (pranavrajs) [inline app/javascript/packs/sdk.js:21]: Can we use `hideMessageBubble` here. Isn't it the same thing?

File: app/javascript/packs/sdk.js [hunk 2]
```diff
@@ -31,6 +32,16 @@ const runSDK = ({ baseUrl, websiteToken }) => {
       IFrameHelper.events.toggleBubble(state);
     },
 
+    showWidget() {
+      document.querySelector('.woot--bubble-holder').style.display = 'block';
```
Comment (pranavrajs) [inline app/javascript/packs/sdk.js:36]: Just have a check if the element exists. If they used hideMessageBubble before, then I guess the element would not be available.

Also, a create a function similar to the below one.

```js
const toggleBubbleVisibility = visibility => {
  // document..
  // window.$chatwoot.hideMessageBubble = visibility;
}
```

**Review by pranavrajs — CHANGES_REQUESTED** [review]
Please the comments below

File: app/javascript/packs/sdk.js (hunk 2 above)
Comment (fayazara) [reply app/javascript/packs/sdk.js:36]: Shouldn't the name be `toggleWidgetVisibility` ? 

Another question - What happens to new incoming messages when the Widget is hidden?

Comment (pranavrajs) [reply app/javascript/packs/sdk.js:36]: Let us disable only the bubble for now. The PR also does that. You can still call window.$chatwoot.toggle() to show the widget.

File: app/javascript/packs/sdk.js (hunk 1 above)
Comment (fayazara) [reply app/javascript/packs/sdk.js]: Resolved in https://github.com/chatwoot/chatwoot/pull/4223/commits/5f7fa536ab1edd2413bb10d54d0b32cdc61b90cf

File: app/javascript/packs/sdk.js (hunk 2 above)
Comment (fayazara) [reply app/javascript/packs/sdk.js]: Resolved in https://github.com/chatwoot/chatwoot/pull/4223/commits/5f7fa536ab1edd2413bb10d54d0b32cdc61b90cf

File: app/javascript/packs/sdk.js [hunk 3]
```diff
@@ -31,6 +31,13 @@ const runSDK = ({ baseUrl, websiteToken }) => {
       IFrameHelper.events.toggleBubble(state);
     },
 
+    toggleBubbleVisibility(visibility) {
+      document.querySelector('.woot--bubble-holder').style.display = visibility
```
Comment (pranavrajs) [inline app/javascript/packs/sdk.js:35]: To be safe let us use here !!visibility and in the below line.

**Review by pranavrajs — CHANGES_REQUESTED** [review]
Please see the comment above.

File: app/javascript/packs/sdk.js (hunk 3 above)
Comment (fayazara) [reply app/javascript/packs/sdk.js:35]: Resolved in https://github.com/chatwoot/chatwoot/pull/4223/commits/dc0b790ed06d1f12e03e88d0bcee4f47096e57bb

**Review by pranavrajs — DISMISSED** [review]

**Review by iamsivin — DISMISSED** [review]

File: app/javascript/packs/sdk.js (hunk 3 above)
Comment (pranavrajs) [inline app/javascript/packs/sdk.js:35]: @fayazara Please test it in the case where the message bubble is hidden already. 

Initialize it with 

```js
window.chatwootSettings = {
  hideMessageBubble: true
}
```

Comment (fayazara) [reply app/javascript/packs/sdk.js:35]: Okay

File: app/javascript/sdk/sdk.js [hunk 4]
```diff
@@ -227,4 +227,8 @@ export const SDK_CSS = `
     width: 400px !important;
  }
 }
+
+.woot-widget-hidden {
```
Comment (pranavrajs) [inline app/javascript/sdk/sdk.js:231]: Let us name it bubble not widget. Until we come up with a better naming pattern, bubble is the button widget is the iFrameHolder

**Review by pranavrajs — CHANGES_REQUESTED** [review]
A small name change, otherwise looks good.

File: app/javascript/sdk/sdk.js (hunk 4 above)
Comment (fayazara) [reply app/javascript/sdk/sdk.js]: Resolved

**Comment by pranavrajs** [comment]
Screenshot 2022-04-01 at 8 47 38 PM


```js
window.toggleBubbleVisibility('hide');
window.toggle('open')
```

@fayazara There is a bug in the implementation.

**Review by pranavrajs — CHANGES_REQUESTED** [review]
Please see the comment regarding the case where the PR breaks

**Comment by fayazara** [comment]
Added a loom with a demo.

https://www.loom.com/share/319dd7653592498b907043a18aff646f

**Review by iamsivin — APPROVED** [review]
@fayazara Tested locally. LGTM ⚡️
https://www.loom.com/share/557b23ed1fa44c40be73d982c7b16f67

@pranavrajs Is it good to merge?

**Comment by pranavrajs** [comment]
@iamsivin @fayazara There was an issue with the close on the widget.

Screenshot 2022-05-06 at 12 53 29 PM


This close button should appear when the bubble is hidden, I've made the relevant changes in https://github.com/chatwoot/chatwoot/pull/4223/commits/bd87a2f19fb481b88955daf42a2d3d6cac0c4593

Can you please test this again?

**Comment by fayazara** [comment]
> Can you please test this again?

Tested locally, working as expected.

**Comment by iamsivin** [comment]
Tested locally looks good to me 👍🏻.

@pranavrajs Found a bug. I'm not sure. Not from this PR. I think it's there in develop branch also. 
When we call `window.$chatwoot.toggle('hide')` the widget will hide and if we call it one more time the widget will pop up even if it is `("hide")` same vice versa. You can see it on the below screencast.

https://user-images.githubusercontent.com/64252451/167108361-a76640ee-76dc-4c18-82fa-0d79c50fa4cf.mov

**Comment by pranavrajs** [comment]
@iamsivin It should have been window.$chatwoot.toggle('close')

**Pull request merged**